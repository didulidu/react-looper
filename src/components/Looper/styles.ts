import styled from "styled-components";

export const StyledGrid = styled.div`
  display: grid;
  width: 750px;
  grid-template-columns: repeat(3, 1fr);

  @media only screen and (max-width: 900px) {
    width: 450px;
  }
`;

export const StyledFooter = styled.div`
  margin-top: 20px;
  display: flex;
  width: 100%;
`;

export const buttonDimension = { width: "40px", height: "40px" };
