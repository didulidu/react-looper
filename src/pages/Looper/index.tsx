import Looper from "./Looper";

export default Looper;
