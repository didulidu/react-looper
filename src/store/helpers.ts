import { Pad } from "../model/Pad";

export const refreshPads = (pads: Pad[]) => {
  return pads.map((pad) => {
    pad.audio.currentTime = 0;
    return {
      ...pad,
      isPlaying: pad.active,
    };
  });
};

export const getPadsFromUrls = (urls: string[]) =>
  urls.map((url) => {
    const audio = new Audio(url);
    return {
      id: url,
      active: false,
      audio,
      isPlaying: !audio.paused,
    };
  });
