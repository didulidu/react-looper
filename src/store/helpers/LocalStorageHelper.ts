import { parseJson } from "../../utils/jsonHelpers";

class LocalStorageHelper {
  static set<T>(key: string | string, value: T): void {
    window.localStorage.setItem(key, JSON.stringify(value));
  }

  static get<T>(key: string | string): T | null {
    const dataString = window.localStorage.getItem(key);
    if (dataString) {
      return parseJson(dataString);
    }
    return null;
  }

  static remove(key: string | string): void {
    window.localStorage.removeItem(key);
  }
}

export default LocalStorageHelper;
