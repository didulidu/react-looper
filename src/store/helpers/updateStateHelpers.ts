import { Pad } from "../../model/Pad";

export const updatePads = (pads: Pad[]) => {
  return pads.map((pad) => {
    pad.audio.currentTime = 0;
    return {
      ...pad,
      isPlaying: pad.active,
    };
  });
};
