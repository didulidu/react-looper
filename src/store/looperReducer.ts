import { Pad } from "../model/Pad";
import { getPadsFromUrls, refreshPads } from "./helpers";

export type LooperStateType = {
  pads: Pad[];
};

export type ActionType =
  | {
      type: "playSingle";
      padId: string;
    }
  | { type: "stopSingle"; pad: Pad }
  | { type: "playAll" }
  | { type: "stopAll" }
  | { type: "startInterval" };

export const getDefaultLooperReducerState = (): LooperStateType => {
  const urls = require.context("../../public", false, /\.(mp3|mpga)$/).keys();
  return {
    pads: getPadsFromUrls(urls),
  };
};

export const looperReducer = (state: LooperStateType, action: ActionType) => {
  switch (action.type) {
    case "playSingle": {
      const updatedPads = state.pads.map((pad) => {
        if (pad.id === action.padId) {
          const updatePad = {
            ...pad,
            active: !pad.active,
            isPlaying: pad.isPlaying && pad.active ? false : pad.isPlaying,
          };
          return updatePad;
        }
        return { ...pad };
      });

      return { ...state, pads: updatedPads };
    }
    case "stopSingle": {
      return state;
    }
    case "playAll": {
      const updatedPads = state.pads.map((pad) => {
        pad.audio.currentTime = 0;
        const updatePad = {
          ...pad,
          active: true,
          isPlaying: true,
        };
        return updatePad;
      });

      return { ...state, pads: updatedPads };
    }
    case "stopAll": {
      const updatedPads = state.pads.map((pad) => {
        const updatePad = {
          ...pad,
          active: false,
          isPlaying: false,
        };
        return updatePad;
      });

      return { ...state, pads: updatedPads };
    }
    case "startInterval": {
      return { ...state, pads: refreshPads(state.pads) };
    }
  }
};
