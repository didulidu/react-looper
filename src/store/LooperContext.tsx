import {
  createContext,
  FC,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useReducer,
  useState,
} from "react";

import { looperReducer, getDefaultLooperReducerState } from "./looperReducer";
import { LooperContextType } from "./types";

export const LooperContext = createContext<LooperContextType>(
  {} as LooperContextType
);

export const useLooperContext = () => useContext(LooperContext);

export const LooperProvider: FC = ({ children }) => {
  const urls = require.context("../../public", false, /\.(mp3|mpga)$/).keys();
  console.log(urls);
  const [intervalId, setIntervalId] = useState<null | NodeJS.Timer>(null);

  const [{ pads }, dispatch] = useReducer(
    looperReducer,
    getDefaultLooperReducerState(urls)
  );

  const isLooperInactive = useMemo(
    () => pads.every((pad) => !pad.active),
    [pads]
  );

  const shouldStartInterval = useMemo(
    () => !isLooperInactive && !intervalId,
    [intervalId, isLooperInactive]
  );

  const resetInterval = useCallback(() => {
    const intervalId = setInterval(() => {
      dispatch({ type: "startInterval" });
    }, 8.01955 * 1000);
    setIntervalId(intervalId);
  }, []);

  useEffect(() => {
    if (shouldStartInterval) {
      dispatch({ type: "startInterval" });
      resetInterval();
    }
  }, [dispatch, shouldStartInterval, resetInterval]);

  useEffect(() => {
    if (isLooperInactive) {
      clearInterval(intervalId!);
      setIntervalId(null);
    }
    return () => clearInterval(intervalId!);
  }, [intervalId, isLooperInactive]);

  const playSingle = useCallback(
    (padId: string) => {
      dispatch({ type: "playSingle", padId: padId });
    },
    [dispatch]
  );

  const playAll = useCallback(() => {
    clearInterval(intervalId!);
    resetInterval();
    dispatch({ type: "playAll" });
  }, [dispatch, intervalId, resetInterval]);

  const stopAll = useCallback(() => {
    dispatch({ type: "stopAll" });
  }, [dispatch]);

  return (
    <LooperContext.Provider value={{ pads, playSingle, playAll, stopAll }}>
      {children}
    </LooperContext.Provider>
  );
};
