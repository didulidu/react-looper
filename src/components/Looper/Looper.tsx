import { FC } from "react";
import { useLooperContext } from "../../store/LooperContext";
import Button from "../atoms/Button/Button";
import Pad from "../atoms/Pad/Pad";
import { StyledFooter, StyledGrid, buttonDimension } from "./styles";
import { ReactComponent as ButtonPlayIcon } from "../../assets/icons/button_play.svg";
import { ReactComponent as ButtonPauseIcon } from "../../assets/icons/button_pause.svg";

const Looper: FC = () => {
  const { pads, playSingle, playAll, stopAll } = useLooperContext();

  return (
    <div>
      <StyledGrid>
        {pads.map((pad) => {
          return <Pad key={pad.id} pad={pad} onPress={playSingle} />;
        })}
      </StyledGrid>
      <StyledFooter>
        <Button onClick={playAll} color="lightblue" aria-label="Play All">
          <ButtonPlayIcon style={buttonDimension} />
        </Button>
        <Button onClick={stopAll} color="indianred" aria-label="Stop All">
          <ButtonPauseIcon style={buttonDimension} />
        </Button>
      </StyledFooter>
    </div>
  );
};

export default Looper;
