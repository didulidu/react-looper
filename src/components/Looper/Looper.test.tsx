import { render, screen } from "@testing-library/react";
import Looper from ".";
import { getPadsFromUrls } from "../../store/helpers";
import { useLooperContext } from "../../store/LooperContext";
import userEvent from "@testing-library/user-event";

jest.mock("../../store/LooperContext", () => ({
  useLooperContext: jest.fn(),
}));

const useLooperContextMock = useLooperContext as jest.Mock;

const playAllMock = jest.fn();
const stopAllMock = jest.fn();
const playSingleMock = jest.fn();

describe("Looper test", () => {
  beforeEach(() => {
    useLooperContextMock.mockImplementation(() => ({
      pads: getPadsFromUrls([
        "./loop1.mp3",
        "./loop2.mp3",
        "./loop3.mpga",
        "./loop4.mpga",
        "./loop5.mp3",
        "./loop6.mp3",
        "./loop7.mp3",
        "./loop8.mp3",
        "./loop9.mp3",
      ]),
      playAll: playAllMock,
      stopAll: stopAllMock,
      playSingle: playSingleMock,
    }));
  });
  test("Looper - renders correctly", () => {
    const { container } = render(<Looper />);
    expect(container).toMatchSnapshot();
  });
  test("Looper - play all", () => {
    render(<Looper />);
    userEvent.click(screen.getByRole("button", { name: "Play All" }));
    expect(playAllMock).toBeCalledTimes(1);
  });
  test("Looper - stop all", () => {
    render(<Looper />);
    userEvent.click(screen.getByRole("button", { name: "Stop All" }));
    expect(stopAllMock).toBeCalledTimes(1);
  });
  test("Looper - play single", () => {
    render(<Looper />);
    userEvent.click(screen.getAllByLabelText("Pad")[0]);
    expect(playSingleMock).toBeCalledTimes(1);
  });
});
