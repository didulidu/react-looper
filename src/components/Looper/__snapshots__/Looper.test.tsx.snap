// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Looper test Looper - renders correctly 1`] = `
<div>
  <div>
    <div
      class="sc-dkPtRN jFSCBy"
    >
      <div
        aria-label="Pad"
        class="sc-gsDKAQ iCqTAi"
      >
        <svg
          height="100px"
          width="100px"
        >
          play.svg
        </svg>
      </div>
      <div
        aria-label="Pad"
        class="sc-gsDKAQ iCqTAi"
      >
        <svg
          height="100px"
          width="100px"
        >
          play.svg
        </svg>
      </div>
      <div
        aria-label="Pad"
        class="sc-gsDKAQ iCqTAi"
      >
        <svg
          height="100px"
          width="100px"
        >
          play.svg
        </svg>
      </div>
      <div
        aria-label="Pad"
        class="sc-gsDKAQ iCqTAi"
      >
        <svg
          height="100px"
          width="100px"
        >
          play.svg
        </svg>
      </div>
      <div
        aria-label="Pad"
        class="sc-gsDKAQ iCqTAi"
      >
        <svg
          height="100px"
          width="100px"
        >
          play.svg
        </svg>
      </div>
      <div
        aria-label="Pad"
        class="sc-gsDKAQ iCqTAi"
      >
        <svg
          height="100px"
          width="100px"
        >
          play.svg
        </svg>
      </div>
      <div
        aria-label="Pad"
        class="sc-gsDKAQ iCqTAi"
      >
        <svg
          height="100px"
          width="100px"
        >
          play.svg
        </svg>
      </div>
      <div
        aria-label="Pad"
        class="sc-gsDKAQ iCqTAi"
      >
        <svg
          height="100px"
          width="100px"
        >
          play.svg
        </svg>
      </div>
      <div
        aria-label="Pad"
        class="sc-gsDKAQ iCqTAi"
      >
        <svg
          height="100px"
          width="100px"
        >
          play.svg
        </svg>
      </div>
    </div>
    <div
      class="sc-hKwDye iaeFjz"
    >
      <button
        aria-label="Play All"
        class="sc-bdvvtL ljBphM"
        role="button"
        type="button"
      >
        <svg
          style="width: 40px; height: 40px;"
        >
          button_play.svg
        </svg>
      </button>
      <button
        aria-label="Stop All"
        class="sc-bdvvtL vXBuu"
        role="button"
        type="button"
      >
        <svg
          style="width: 40px; height: 40px;"
        >
          button_pause.svg
        </svg>
      </button>
    </div>
  </div>
</div>
`;
